### ISSUE TITLE
Taskbar mods not loading (`ERROR_INTERNET_CANNOT_CONNECT`)

### ISSUE THREAD

**Issue by nazzy121** [ramensoftware/windhawk-mods#701]
Any workaround for this?  Tried copying symbols from another PC (where mods load successfully with identical Windows build (22631.3593), but still the mods won't load.  

![image](https://github.com/ramensoftware/windhawk-mods/assets/158132428/6a42d4e5-586d-4bc2-a803-f76ea8411333)

![image](https://github.com/ramensoftware/windhawk-mods/assets/158132428/05d9098b-28c5-42fc-8191-81f0720a9f32)

[DebugViewConsole 1.8.0.102.txt](https://github.com/ramensoftware/windhawk-mods/files/15478239/DebugViewConsole.1.8.0.102.txt)

**Comment by m417z** [comment]
The error in the log is `ERROR_INTERNET_CANNOT_CONNECT`. Do you have a firewall or an antivirus? Try to exclude explorer.exe, then disable and re-enable the mod.

Basically the problem is that explorer.exe isn't able to download symbols from the internet, so there's something with your computer setup.

Alternatively, as a workaround, you can try using [Windhawk Symbol Helper](https://github.com/ramensoftware/windhawk-symbol-helper/releases/latest). Run it and set "Target executable" (see paths below), then click "Get symbols". After you did that for all the paths in the list, disable and re-enable the mods. If it will work, note that you'll need to do it after every Windows update.

The paths to set in "Target executable":
```
C:\Windows\explorer.exe
```
```
C:\Windows\System32\Taskbar.dll
```
```
C:\Windows\SystemApps\MicrosoftWindows.Client.Core_cw5n1h2txyewy\Taskbar.View.dll
```

**Comment by nazzy121** [comment]
I have two computers, let's call them PC1 and PC2.  Both are exact same windows version, and are behind the same Netgear router.  I confirmed internet works on both of them.  No antivirus except for built-in Windows Defender on Windows 11.  

Everything works on PC1.   But for some reason, taskbar mods won't load on PC2.  

Question, I copied symbols from PC1 to PC2 expecting to fix the problem, but it didn't.  Should it have?  

I excluded explorer.exe in Windows Defender, now when I enable mods, I get 2 entries of "unloaded" for each mod, instead of 1 entry for each:  
![image](https://github.com/ramensoftware/windhawk-mods/assets/158132428/3383489c-702f-4d46-84fc-ff0e03ab32de)

**Comment by nazzy121** [comment]
I tried [Windhawk Symbol Helper](https://github.com/ramensoftware/windhawk-symbol-helper/releases/latest) but it doesn't seem to work either.  I assume you meant to set "Target executable" to each of those paths 1 by 1.  When i click "Get Symbols" for each target path, it returns this:  Error: symbol_enum.cpp(305)\windhawk-symbol-helper.exe!00007FF6DB97D3FE: (caller: 00007FF6DB974074) Exception(14) tid(a2c) 806D0005 

![image](https://github.com/ramensoftware/windhawk-mods/assets/158132428/12a4237b-82c9-4181-a768-eb08c6fdd450)

![image](https://github.com/ramensoftware/windhawk-mods/assets/158132428/9628639d-178e-48e4-a42c-176f9e885646)

![image](https://github.com/ramensoftware/windhawk-mods/assets/158132428/3404a5ee-329a-48a4-bb20-281b5f573ea4)


Tried copy/pasting Microsoft symbol server address to a web browser:  https://msdl.microsoft.com/download/symbols
returns this error page:

![image](https://github.com/ramensoftware/windhawk-mods/assets/158132428/c03d887b-a7bf-41a8-adbf-e44af0fde86c)

**Comment by m417z** [comment]
> I assume you meant to set "Target executable" to each of those paths 1 by 1.

Yes.

> I tried [Windhawk Symbol Helper](https://github.com/ramensoftware/windhawk-symbol-helper/releases/latest) but it doesn't seem to work either.

Can you check the Log checkbox and post the full log content for one of them?

> Tried copy/pasting Microsoft symbol server address to a web browser: https://msdl.microsoft.com/download/symbols returns this error page:

That's fine. Try a link to a symbol file, for example:
https://msdl.microsoft.com/download/symbols/explorer.pdb/0B226550FBD4A4AD0D11C794496F41811/explorer.pdb

**Comment by nazzy121** [comment]
I put the log next to a ping test (yahoo.com) to prove internet is working :) 
![image](https://github.com/ramensoftware/windhawk-mods/assets/158132428/eaa6e471-5503-4bc1-b3fe-b7c53067ddc7)

Clicking symbol file results in successful download: 
![image](https://github.com/ramensoftware/windhawk-mods/assets/158132428/b69e4197-9d55-4ae1-a331-41a48d6b821a)

**Comment by m417z** [comment]
I'm using Microsoft's SymSrv library. Not sure why it can't connect to the internet on your computer. Maybe something related to proxy settings, although [it claims it supports integrated support for a proxy server](https://learn.microsoft.com/en-us/windows-hardware/drivers/debugger/firewalls-and-proxy-servers).

As another workaround you can do the following:
* Take the full path from the logs, in your case:
```
C:\ProgramData\Windhawk\Engine\Symbols\explorer.pdb\0B226550FBD4A4AD0D11C794496F41811\explorer.pdb
```
* Convert it to a download URL (always starts with `https://msdl.microsoft.com/download/symbols`):
```
https://msdl.microsoft.com/download/symbols/explorer.pdb/0B226550FBD4A4AD0D11C794496F41811/explorer.pdb
```
Download the file, rename it to `explorer.pdb`, and place it to the path from the logs.

Alternatively, just run the following command in cmd:
```
curl --create-dirs --output C:\ProgramData\Windhawk\Engine\Symbols\explorer.pdb\0B226550FBD4A4AD0D11C794496F41811\explorer.pdb https://msdl.microsoft.com/download/symbols/explorer.pdb/0B226550FBD4A4AD0D11C794496F41811/explorer.pdb
```

Do these steps for all three modules (explorer.exe, Taskbar.dll, Taskbar.View.dll).

**Comment by m417z** [comment]
If you have a way to reproduce the problem, please let me know and I'll think of a solution.

Without additional information, all I can provide is a workaround, so I'm closing this issue. If you need more help or have more questions, feel free to follow up.

**Comment by nazzy121** [comment]
How do you get this number: "0B226550FBD4A4AD0D11C794496F41811" ?  Is it the same number for Taskbar.dll and Taskbar.View.dll?

**Comment by m417z** [comment]
From the logs. It's different for each module.

![image](https://github.com/ramensoftware/windhawk-mods/assets/4129781/40515cc3-c0b4-41a3-aedd-5987df2275c1)

**Comment by nazzy121** [comment]
Unfortunately I get a 0kb file downloaded :( 

![image](https://github.com/ramensoftware/windhawk-mods/assets/158132428/3ea726e6-1d65-4a06-ab89-af9017eca537)

**Comment by m417z** [comment]
It's because redirections aren't followed. My bad, sorry. Add `-L` after `curl --create-dirs` and try again, like this:

```
curl --create-dirs -L --output C:\ProgramData\Windhawk\Engine\Symbols\explorer.pdb\0B226550FBD4A4AD0D11C794496F41811\explorer.pdb https://msdl.microsoft.com/download/symbols/explorer.pdb/0B226550FBD4A4AD0D11C794496F41811/explorer.pdb
```

**Comment by nazzy121** [comment]
![image](https://github.com/ramensoftware/windhawk-mods/assets/158132428/93b5e437-c0ec-4ac9-a282-f8c81c45e530)

Success :)   So do I have to run these commands after every windows update?  And do the long numbers also change?

**Comment by m417z** [comment]
> So do I have to run these commands after every windows update? And do the long numbers also change?

Yes and yes.

> Windhawk Symbol Helper is now suddenly working.... so maybe the command lines somehow fixed the connection?

It's working because it uses the local files you just downloaded.

**Comment by nazzy121** [comment]
I wonder if it's a DNS issue.  So I deleted the .pdb files to test a different DNS server, but the mods still load (even after reboot).  

Is there a way to revert back to when the mods could not load?

**Comment by m417z** [comment]
Maybe try running the curl command with verbose logs: `curl -v ...`

And try running some standard project that uses WinInet, like this one (I suspect it won't work):
https://www.codeproject.com/Articles/25185/WinINet-Test-Application

And see if there's anything anywhere in the logs that provides a hint for the reason of the failure.

**Comment by m417z** [comment]
Here, I compiled the WinInetTest tool:
[WinInetTest.zip](https://github.com/user-attachments/files/15521086/WinInetTest.zip)

**Comment by nazzy121** [comment]
![image](https://github.com/ramensoftware/windhawk-mods/assets/158132428/b8d4b188-3794-47ff-80bc-8fdf03b43c39)



[curl-v.txt](https://github.com/user-attachments/files/15521103/curl-v.txt)

**Comment by m417z** [comment]
Error 12029 is `ERROR_INTERNET_CANNOT_CONNECT`, quite generic unfortunately.
What if you uncheck "Use SSL"?

**Comment by nazzy121** [comment]
> What if you uncheck "Use SSL"?

No difference unfortunately.   

![image](https://github.com/ramensoftware/windhawk-mods/assets/158132428/44d103a6-efde-41a9-ae1c-fa3c3713e90c)

**Comment by nazzy121** [comment]
This is so weird.  I can connect to everything else.  Websites, game servers, cloud apps like Dropbox, corporate VPN, etc.

**Comment by m417z** [comment]
Do you have a proxy server configured? You can check with:
```
reg query "HKCU\Software\Microsoft\Windows\CurrentVersion\Internet Settings" | find /i "proxy"
```

**Comment by nazzy121** [comment]
No, it's pretty much a bare W11 install with some apps including VMware Workstation and a vpn client.  

![image](https://github.com/ramensoftware/windhawk-mods/assets/158132428/ffb47576-e5b5-46d9-b8fb-04b0a179806c)

**Comment by m417z** [comment]
Then I'm afraid I'm out of ideas. You can Google for "HttpSendRequest ERROR_INTERNET_CANNOT_CONNECT" and look for more solutions. At least we verified that it's not Windhawk alone. If you find anything, let me know.

**Comment by nazzy121** [comment]
Think I found it :)  I remembered Internet Options had proxy configurations so I went there to double check that no proxy was set up (there wasn't any).  Then I looked at "LAN Settings" -> Automatically detect settings was NOT checkmarked.    I compared this to my working PC which was checkmarked.  So I just click the checkbox on problem PC, then WinInitTest worked.   I don't know why it was unchecked, I don't remember ever touching that setting.  Anyway, I'm fairly sure this was causing the issue.  

![image](https://github.com/ramensoftware/windhawk-mods/assets/158132428/70ef3c21-9e52-413c-9c36-3cd974ad64b4)
